### PR TITLE
Generate vaunt file

### PR DESCRIPTION
**Pull request by Felipe444**

### CODE REVIEW

File: vaunt-generator/src/test/groovy/com/hltech/vaunt/generator/domain/representation/RepresentationExtractorSpec.groovy [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            service.capabilities.contracts[0].destinationType == DestinationType.QUEUE
+            service.capabilities.contracts[0].body == jsonSchemaGenerator.generateSchema(SampleProviderMessage.class)
+
+        and: 'Expectations are as expected'
+            service.expectations
+            service.expectations.providerNameToContracts
+            service.expectations.providerNameToContracts.size() == 1
+            service.expectations.providerNameToContracts.get('provider')
+            service.expectations.providerNameToContracts.get('provider').size() == 1
+            def providerContract = service.expectations.providerNameToContracts.get('provider')[0]
```
Comment (matek2305) [inline vaunt-generator/src/test/groovy/com/hltech/vaunt/generator/domain/representation/RepresentationExtractorSpec.groovy:44]: i would like to add separate `and:` block for contract here, what do you think?

File: vaunt-generator/src/test/groovy/com/hltech/vaunt/generator/VauntGeneratorSpec.groovy [hunk 2]
```diff
@@ -2,11 +2,11 @@ package com.hltech.vaunt.generator
 
 import spock.lang.Specification
 
-class AppSpec extends Specification {
+class VauntGeneratorSpec extends Specification {
```
Comment (matek2305) [inline vaunt-generator/src/test/groovy/com/hltech/vaunt/generator/VauntGeneratorSpec.groovy:5]: should it be removed?

File: vaunt-generator/src/main/java/com/hltech/vaunt/generator/domain/representation/RepresentationExtractor.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+    private final JsonSchemaGenerator jsonSchemaGenerator;
+
+    public Service extractServiceRepresentation(String packageRoot, String serviceName) throws JsonMappingException {
+        return new Service(serviceName, extractCapabilities(packageRoot), extractExpectations(packageRoot));
+    }
+
+    private Capabilities extractCapabilities(String packageRoot) throws JsonMappingException {
+        List<Contract> providerContracts = new ArrayList<>();
+        for (Class<?> providerMessage : new Reflections(packageRoot).getTypesAnnotatedWith(Provider.class)) {
```
Comment (matek2305) [inline vaunt-generator/src/main/java/com/hltech/vaunt/generator/domain/representation/RepresentationExtractor.java:30]: why not use streams here?

File: vaunt-generator/src/main/java/com/hltech/vaunt/generator/domain/representation/RepresentationExtractor.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+                    providerMessage.getAnnotation(Provider.class).destinationName(),
+                    jsonSchemaGenerator.generateSchema(providerMessage)));
+        }
+
+        return new Capabilities(providerContracts);
+    }
+
+    private Expectations extractExpectations(String packageRoot) throws JsonMappingException {
+        Multimap<String, Contract> providerNameToContracts = ArrayListMultimap.create();
+        for (Class<?> consumerMessage : new Reflections(packageRoot).getTypesAnnotatedWith(Consumer.class)) {
```
Comment (matek2305) [inline vaunt-generator/src/main/java/com/hltech/vaunt/generator/domain/representation/RepresentationExtractor.java:42]: also here

**Review by matek2305 — CHANGES_REQUESTED** [review]

File: vaunt-generator/src/test/groovy/com/hltech/vaunt/generator/VauntGeneratorSpec.groovy (hunk 2 above)
Comment (Felipe444) [reply vaunt-generator/src/test/groovy/com/hltech/vaunt/generator/VauntGeneratorSpec.groovy:5]: Implemented later.

File: vaunt-generator/src/main/java/com/hltech/vaunt/generator/domain/representation/RepresentationExtractor.java (hunk 3 above)
Comment (Felipe444) [reply vaunt-generator/src/main/java/com/hltech/vaunt/generator/domain/representation/RepresentationExtractor.java:30]: For me in this particular case current approach is more readable.

File: vaunt-generator/src/main/java/com/hltech/vaunt/generator/domain/representation/RepresentationExtractor.java (hunk 4 above)
Comment (Felipe444) [reply vaunt-generator/src/main/java/com/hltech/vaunt/generator/domain/representation/RepresentationExtractor.java:42]: same

File: vaunt-generator/src/test/groovy/com/hltech/vaunt/generator/domain/representation/RepresentationExtractorSpec.groovy (hunk 1 above)
Comment (Felipe444) [reply vaunt-generator/src/test/groovy/com/hltech/vaunt/generator/domain/representation/RepresentationExtractorSpec.groovy:44]: it is a part of expectations

File: vaunt-generator/src/main/java/com/hltech/vaunt/generator/domain/representation/RepresentationExtractor.java (hunk 3 above)
Comment (matek2305) [reply vaunt-generator/src/main/java/com/hltech/vaunt/generator/domain/representation/RepresentationExtractor.java:30]:
```java
List<Contract> providerContracts = new Reflections(packageRoot).getTypesAnnotatedWith(Provider.class)
    .stream()
    .map(toContract())
    .collect(Collectors.toList());

return new Capabilities(providerContracts);
```

this is quite nice for readability and "first-look" complexity in my opinion

File: vaunt-generator/src/test/groovy/com/hltech/vaunt/generator/VauntGeneratorSpec.groovy (hunk 2 above)
Comment (matek2305) [reply vaunt-generator/src/test/groovy/com/hltech/vaunt/generator/VauntGeneratorSpec.groovy:5]: ok but i do not understand why to keep this file in repo if it is not implemented

**Review by matek2305 — CHANGES_REQUESTED** [review]

File: vaunt-generator/src/test/groovy/com/hltech/vaunt/generator/VauntGeneratorSpec.groovy (hunk 2 above)
Comment (Felipe444) [reply vaunt-generator/src/test/groovy/com/hltech/vaunt/generator/VauntGeneratorSpec.groovy:5]: yes, I removed it :). It shold be implemented later as I wrote above.

File: vaunt-generator/src/main/java/com/hltech/vaunt/generator/domain/representation/RepresentationExtractor.java (hunk 3 above)
Comment (Felipe444) [reply vaunt-generator/src/main/java/com/hltech/vaunt/generator/domain/representation/RepresentationExtractor.java:30]: It is not that easy.
1. For now we have no approach to handle exceptions (it must be decided separately)
2. jsonSchemaGenerator.generateSchema(...) throws an exception. It means more complicated implementation than suggested by you -> we would need to handle it in toContract(...) method
3. still I prefer earlier approach

Comment (Felipe444) [reply vaunt-generator/src/main/java/com/hltech/vaunt/generator/domain/representation/RepresentationExtractor.java]: Alternative is something similar to:


```suggestion
    private Capabilities extractCapabilities(String packageRoot) {
        List<Contract> providerContracts = new Reflections(packageRoot).getTypesAnnotatedWith(Provider.class).stream()
                .map(this::toContract)
                .collect(Collectors.toList());

        return new Capabilities(providerContracts);
    }

    private Contract toContract(Class<?> providerMessage) {
        JsonSchema body = null;
        try {
            body = jsonSchemaGenerator.generateSchema(providerMessage);
        } catch (JsonMappingException e) {
            e.printStackTrace();
        }
        return new Contract(
                providerMessage.getAnnotation(Provider.class).destinationType(),
                providerMessage.getAnnotation(Provider.class).destinationName(),
                body);
    }
```

Comment (Felipe444) [reply vaunt-generator/src/main/java/com/hltech/vaunt/generator/domain/representation/RepresentationExtractor.java]: Alternative would be similar to:

`    private Capabilities extractCapabilities(String packageRoot) {
        List<Contract> providerContracts = new Reflections(packageRoot).getTypesAnnotatedWith(Provider.class).stream()
                .map(this::toContract)
                .collect(Collectors.toList());

        return new Capabilities(providerContracts);
    }

    private Contract toContract(Class<?> providerMessage) {
        JsonSchema body = null;
        try {
            body = jsonSchemaGenerator.generateSchema(providerMessage);
        } catch (JsonMappingException e) {
            e.printStackTrace();
        }
        return new Contract(
                providerMessage.getAnnotation(Provider.class).destinationType(),
                providerMessage.getAnnotation(Provider.class).destinationName(),
                body);
    }`

Comment (Felipe444) [reply vaunt-generator/src/main/java/com/hltech/vaunt/generator/domain/representation/RepresentationExtractor.java:30]: An alternative would be somthing similar to:

```
   private Capabilities extractCapabilities(String packageRoot) {
        List<Contract> providerContracts = new Reflections(packageRoot).getTypesAnnotatedWith(Provider.class).stream()
                .map(this::toContract)
                .collect(Collectors.toList());

        return new Capabilities(providerContracts);
   }

   private Contract toContract(Class<?> providerMessage) {
        JsonSchema body = null;
        try {
            body = jsonSchemaGenerator.generateSchema(providerMessage);
        } catch (JsonMappingException e) {
            e.printStackTrace();
        }
        return new Contract(
                providerMessage.getAnnotation(Provider.class).destinationType(),
                providerMessage.getAnnotation(Provider.class).destinationName(),
                body);
    }
```

**Review by matek2305 — APPROVED** [review]

**Pull request merged**